src/skydb_internal.ts

```diff
@@ -0,0 +1,192 @@
+  const { data } = await client.db.getRawBytes(publicKey, dataKey, opts);
+  if (data === null) {
+    return { data: null };
+  }
+
+  const encryptionKey = deriveEncryptedFileKeyEntropy(pathSeed);
+  const json = decryptJSONFile(data, encryptionKey);
+
+  return { data: json };
+}
+
+/**
+ * Sets Encrypted JSON at the given path through MySky.
+ *
+ * @param client - The Skynet client.
+ * @param seed - The user seed.
+ * @param path - The data path.
+ * @param json - The json to encrypt and set.
+ * @returns - An object containing the original json data.
+ * @throws - Will throw if the user does not have Hidden Write permission on the path.
+ */
+export async function setJSONEncryptedInternal(
+  client: SkynetClient,
+  seed: Uint8Array,
+  path: string,
+  json: JsonData
+): Promise<EncryptedJSONResponse> {
+  log("Entered setJSONEncryptedInternal");
+
+  validateString("path", path, "parameter");
+  validateObject("json", json, "parameter");
+
+  const { publicKey } = genKeyPairFromSeed(seed);
+  const pathSeed = await getEncryptedPathSeedInternal(seed, path, false);
+  const dataKey = deriveEncryptedFileTweak(pathSeed);
+  const opts = { hashedDataKeyHex: true };
+
+  // Immediately fail if the mutex is not available.
+  return await client.db.revisionNumberCache.withCachedEntryLock(publicKey, dataKey, async (cachedRevisionEntry) => {
+    // Get the cached revision number before doing anything else.
+    const newRevision = incrementRevision(cachedRevisionEntry.revision);
+
+    // Derive the key.
+    const encryptionKey = deriveEncryptedFileKeyEntropy(pathSeed);
+
+    // Pad and encrypt json file.
+    log("Calling encryptJSONFile");
+    const data = encryptJSONFile(json, { version: ENCRYPTED_JSON_RESPONSE_VERSION }, encryptionKey);
+
+    log("Calling getOrCreateSkyDBRegistryEntry");
+    const [entry] = await getOrCreateSkyDBRegistryEntry(client, dataKey, data, newRevision, opts);
+
+    // Sign the entry.
+    log("Calling signEncryptedRegistryEntryInternal");
+    const signature = await signEncryptedRegistryEntryInternal(seed, entry, path);
+
+    log("Calling postSignedEntry");
+    await client.registry.postSignedEntry(publicKey, entry, signature, opts);
+
+    return { data: json };
+  });
+}
+
+/**
+ * Gets the encrypted path seed for the given path without requiring
+ * permissions. This should NOT be exported - for internal use only.
+ *
+ * @param seed - The user seed.
+ * @param path - The given file or directory path.
+ * @param isDirectory - Whether the path corresponds to a directory.
+ * @returns - The hex-encoded encrypted path seed.
+ */
+async function getEncryptedPathSeedInternal(seed: Uint8Array, path: string, isDirectory: boolean): Promise<string> {
+  log("Entered getEncryptedPathSeedInternal");
+
+  // Compute the root path seed.
+  const rootPathSeedBytes = deriveRootPathSeed(seed);
+
+  // Compute the child path seed.
+  return deriveEncryptedPathSeedForRoot(rootPathSeedBytes, path, isDirectory);
```

Comment:
I see your point, as `deriveEncryptedPathSeedForRoot` here can only be used with the root seed. I would like to keep it this way though so that `deriveEncryptedPathSeedForRoot` is similar to `deriveEncryptedPathSeed` in skynet-js with minimal changes. `skynet-js` contains the reference implementation and I think it's more clear if we try to stick to it with as little changes as possible.